File: LeetCode/Tree/BST/Leetcode_2476_closest_nodes_queries_in_a_binary_search_tree/Leetcode_2476_closest_nodes_queries_in_a_binary_search_tree.cpp
/**
 * @brief: Leetcode_2476_二叉搜索树最近节点查询
 * @link: https://leetcode.cn/problems/closest-nodes-queries-in-a-binary-search-tree/description/
 * @author: liangj.zhang
 * @date: 15/12/2025
 * 
 * @Difficulty: Medium
 * 
 * @Label: binary search & BST
 * 
 * @Retrospect(worthy 1 - 5): 4
 * 
 * @thoughts:
 *  + 【思路 1】：Binary Search + BST
 *      先 inOrder 的方式遍历一下搜索二叉树（BST），得到一个升序的序列
 *      题目要求得到，<= target 的最大值，和 >= target 的最小值；
 *      + >= target 的最小值 -> 即 lower_bound
 *      + <= target, 有了 lower_bound 返回的位置 pos，pos - 1 就是该答案；
 *      
 *      + 分析：
 *          + 时间复杂度：inOrder(O(n)) + q * O(logn)
 *          + 空间复杂度：inOrder(O(n)) 没说平衡二叉树 + 2 * O(q)
 * 
 *      + rank: 
 *          + 写法 1：
 *              + 时间效率：107 ms, 击败 21.63%
 *              + 空间效率：194.14 MB, 击败 17.29%
 *          + 写法 2：
 *              + 时间效率：55 ms, 击败 99.49%
 *              + 空间效率：186.51 MB, 击败 83.97%
 */

#include "../../../../Include/Leetcode/Tree/Tree.h"
using Leetcode::Tree::BinaryTree::TreeNode;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include <vector>
using std::vector;

// 【思路 1 -- 写法 2】
class Solution {
private:
    static vector<int> rec;

    void in_order(TreeNode* root) {
        
        if (!root) return;
        in_order(root->left);
        rec.push_back(root->val);
        in_order(root->right);
    }

public:
    vector<vector<int>> closestNodes(TreeNode* root, vector<int>& queries) {
        
        rec.clear();
        in_order(root);
        int size = rec.size();
        int lower, upper;
        vector<vector<int>> ret{};

        for (auto e : queries) {

            int l{0}, r{size - 1};
            while (l <= r) {
                int mid = l + (r - l) / 2;
                if (rec[mid] < e)
                    l = mid + 1;
                else
                    r = mid - 1;
            }
            lower = (l >= 0 && l < size) ? rec[l] : -1;

            if (lower == e)
                upper = e;
            else {
                int r = l - 1;
                upper = (r >= 0 && r < size) ? rec[r] : -1;
            }
            
            ret.push_back({upper, lower});
        }

        return ret;
    }
};

// 【思路 1 -- 写法 1】
/*
class Solution {
private:
    static vector<int> rec;
    static vector<vector<int>> ret;

    // return 第一个 >= target 的位置
    int lower_bound(int target) {

        int l{0}, r(rec.size() - 1);
        while (l <= r) {

            int mid = l + (r - l) / 2;
            if (rec[mid] < target)
                l = mid + 1;
            else
                r = mid - 1;
        }

        return l;
    }

    void in_order(TreeNode* root) {
        
        if (!root) return;
        in_order(root->left);
        rec.push_back(root->val);
        in_order(root->right);
    }

public:
    vector<vector<int>> closestNodes(TreeNode* root, vector<int>& queries) {
        
        rec.clear();
        ret.clear();
        in_order(root);
        int size = rec.size();
        int lower, upper;

        for (auto e : queries) {

            int l = lower_bound(e);
            lower = (l >= 0 && l < size) ? rec[l] : -1;

            if (lower == e)
                upper = e;
            else {
                int r = l - 1;
                upper = (r >= 0 && r < size) ? rec[r] : -1;
            }
            
            ret.push_back({upper, lower});
        }

        return ret;
    }
};
*/
vector<int> Solution::rec{};
vector<vector<int>> Solution::ret{};

File: LeetCode_Python/Tree/BST/Leetcode_2476_closest_nodes_queries_in_a_binary_search_tree/Leetcode_2476_closest_nodes_queries_in_a_binary_search_tree.py
"""
@brief: Leetcode_2476_二叉搜索树最近节点查询
@link: https://leetcode.cn/problems/closest-nodes-queries-in-a-binary-search-tree/description/
@author: liangj.zhang
@date: 15/12/2025

@Difficulty: Medium

@Label: binary search & BST

@Retrospect(worthy 1 - 5): 4

@thoughts:
  + 【思路 1】：Binary Search + BST
        先用 inOrder 的方式遍历二叉搜索树（BST），得到一个升序序列；
        题目要求得到：
            - <= target 的最大值
            - >= target 的最小值

        其中：
            + >= target 的最小值：即 lower_bound
            + <= target：已知 lower_bound 返回的位置 pos，pos - 1 即为答案

        分析：
            时间复杂度：inOrder(O(n)) + q * O(logn)
            空间复杂度：inOrder(O(n))（未假设平衡 BST） + O(q)

        rank:
            时间效率：418 ms, 击败 24.00%
            空间效率：73.6 MB, 击败 18.53%
"""

import os
import sys

print(os.path.join(os.path.dirname(__file__)))
print(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../.."))) #f:\Code\zljgit\LeetCode
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../..")))

from typing import Optional, List

from Include.PyLeetcode.Tree.Tree import TreeNode

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def __init__(self) -> None:
        self.arr = []
        self.ret = []

    def dfs(self, root: Optional[TreeNode]) -> None:
        if not root: return
        self.dfs(root.left)
        self.arr.append(root.val)
        self.dfs(root.right)
        
    def lower_bound(self, target: int) -> int:      
        l, r = 0, self.arr.__len__() - 1
        while l <= r:
            mid = (l + r) // 2
            if self.arr[mid] < target:
                l = mid + 1
            else:
                r = mid - 1
        return l

    def closestNodes(self, root: Optional[TreeNode], queries: List[int]) -> List[List[int]]:

        self.dfs(root)

        size = self.arr.__len__()
        for e in queries:
            l = self.lower_bound(e)
            _min = self.arr[l] if l >= 0 and l < size else -1
            r = l - 1
            _max = e if _min == e else self.arr[r] if r >= 0 and r < size else -1
            self.ret.append([_max, _min])

        return self.ret